### ISSUE TITLE
Organising the conda communities and establishing best practices. 

### ISSUE THREAD

**Issue by bgruening** [conda-forge/staged-recipes#299]
We all love conda and there are many communities that build awesome packages that are easy to use. I would like to see more exchange between these communities to finally share more build-scripts, to develop one best-practice guide and finally to have channels that can be used together without breaking recipes - a list of trusted channels with similar guidelines.

For example the [bioconda](https://github.com/bioconda/bioconda-recipes) community - specialised on bioinformatic software. They have some very nice guides how to develop packages, they are reviewing and bulk-patches recipes if there are new features in conda to make the overall experience even better.
ping @johanneskoester, @daler and @chapmanb from BioConda fame

[Omnia](https://github.com/omnia-md/conda-recipes) has a lot of cheminformatic software and a nice build-box based on phusion/holy-build-box-64 + CUDA and AMD APP SDK.
ping @kyleabeauchamp, @jchodera

With conda-forge there is now a new one and it would be great to get all interested people together to join forces here and don't replicate our recipes or copy them from one channel to the other just to make them compatible.

Another point is that we probably want to move recipes to `default` at some point and deliver our work back to Continuum - so that we can benefit from each other.

I can imagine that we all form a group of trusted communities and channels and activate them by default in our unified build-box - or we have one giant community channel. All this I would like to discuss with everyone that is interested and come up with a plan how to make this happen :)

What do you all think about this?
As a next step I would like to create a doodle to find a meeting data where at least one representative from all communities can participate.

Many thanks to Continuum Analytics for there continues support and the awesome development behind scientific python and this package manager.
ping @jakirkham @msarahan

**Comment by bgruening** [comment]
@kyleabeauchamp, @jchodera, @jakirkham @msarahan, @johanneskoester, @daler, @chapmanb, @jxtx, @jmchilton: please feel free to ping others and invite them :)

**Comment by jchodera** [comment]
Definitely interested in learning more!  For now, pinging @rmcgibbo, @mpharrigan, @cxhernandez, @marscher, @franknoe, @pgrinaway, @bas-rustenburg.

**Comment by ocefpaf** [comment]
@bgruening thanks for the message! In fact we just discussed that yesterday!!

Conda-forge was born from two communities similar to bioconda and omnia (the SciTools and IOOS channels)  with the goal to reduce redundancy and join forces to produce high quality recipes and binaries.

I would love to see more communities join us here. We are not the dark side but we do have cookies :wink: 

I am trying to put a blog post online next week with more info. We are also planning on public (Google?) hangouts so we can have some online face-time and answer questions.

Meanwhile feel free to ask anything here, or in new issues if the you have a very specific question.

Here is the gist of conda-forge:

- We are community driven conda packing loving geeks.
- The entry step is this repository (staged-recipes) people must add their recipes here via PRs.
- Once the PR is merged a feedstock (an individual repo) is created for that recipe.
- Each feedstock has a team of maintainers. The maintainers are specified in the recipe via the `extra/maintainers` field. (Note that feedstock creation and the GitHub happens automagically.)
- We have a cookie cutter/do-it-all tool called [`conda-smithy`](https://github.com/conda-forge/conda-smithy). We use that to lint the recipes, update the feedstocks, and some convinence tools to work with the many repos model.
- A maintainer "job" is mostly to update the software version and merge an eventual auto mantiance PR (like CI config updates).

There are many details I am leaving out and much more to talk about, but I will stop here for now. The number one question we get is: why multiple repo instead of 1 repo with all the recipes. We had (and still have) many discussion like [this](https://github.com/conda-forge/conda-forge.github.io/issues/25). However, all I have to say is this. Why try the single repo model and now we are trying the multiple repo model. So far the multiple repo has scaled much better, and none of the major worries we had became true.

**Comment by jchodera** [comment]
This sounds great. @rmcgibbo is much more qualified to comment than I am here---he pioneered most of the [omnia](http://omnia-md.org) conda framework---but we ended up converging on our own build system (modeled loosely on [`conda/conda-recipes`](https://github.com/conda/conda-recipes) simply because we weren't aware of any other way to tackle this.

Where should we look for all the gory technical details about the build systems and automation? This was the hardest part for us, since we needed (1) broad platform support (hence the use of a [phusion/holy-build-box-64](http://phusion.github.io/holy-build-box/) build system for `linux`, (2) CUDA and OpenCL support (via the AMD APP SDK), and (3) automated builds in reproducible environments for `win`, `linux`, and `osx`. We're also trying to keep old versions of packages live for scientific reproducibility---we frequently publish code with our papers and provide [`environment.yml`](http://conda.pydata.org/docs/using/envs.html#share-an-environment) files to ensure reproducibility with identical versions. Our approach started with actual local hardware and evolved to use cloud services (currently [`travis-ci`](https://travis-ci.org/) and [AppVeyor](http://www.appveyor.com/).  

I'd love to understand more about how the `conda-forge` build system differs from what we currently use in [omnia's build system](https://github.com/omnia-md/conda-recipes).

**Comment by jakirkham** [comment]
> We are not the dark side but we do have cookies :wink:

Which ones? For humans or browsers? :laughing: Ok, it was terrible, but I had no self-control.

<br>
Yes, welcome all. :smile:

Please feel free to peruse what is going on at conda-forge and ask questions. The best place to get acquainted with or propose general discussion topics is probably the [website repo]( https://github.com/conda-forge/conda-forge.github.io ) (in particular the issue tracker). There are many issues there that are likely of interest and welcome to healthy discussion of thoughts and personal experiences. Also, there may be a few closed issues there worth reading up on just to get a little bit of history (we are still quiet young :wink:).

If you would like, feel free to submit a simple recipe or a few to get a feel for how everything works here. Also, feel free to check out our [gitter channel]( https://gitter.im/conda-forge/conda-forge.github.io ) for any generic questions or you may have.

Once everyone has had a chance to get a feel for how everything works and what seems personally relevant, we can figure out meeting discussion topics in some place YTBD.

Again welcome.

**Comment by jakirkham** [comment]
> Where should we look for all the gory technical details about the build systems and automation?

This varies depending on the question. Let's try and direct you based on the points raised.

> (1) broad platform support (hence the use of a [phusion/holy-build-box-64](http://phusion.github.io/holy-build-box/) build system for `linux`

This [issue]( https://github.com/conda-forge/conda-forge.github.io/issues/29 ) has basically moved in the direction of various proposals of how to move the Linux build system forward. Though there is current strategy in place, as well.

> (2) CUDA and OpenCL support (via the AMD APP SDK)...

This is under active discussion. The reason being this is tied to several issues including build system constraints, how features work, how and what of these libraries get distributed, etc.. See this [issue]( https://github.com/conda-forge/conda-forge.github.io/issues/63 ). There is a proposed example there of how we might get this to work. However, we haven't settled on something yet. 

> (3) automated builds in reproducible environments for `win`, `linux`, and `osx`.

This is all over the map. :smile: In general, we use AppVeyor (Windows), Travis CI (Mac), and Circle CI (Dockerized Linux)

If you just want to read code, we can point you there. Proper documentation isn't quite there yet. Also, there isn't one singular issue for this, but it is discussed at various points in various issues. What sort of things would you like to know?

**Comment by jakirkham** [comment]
Welcome @jchodera.

> Where should we look for all the gory technical details about the build systems and automation?

This varies depending on the question. Let's try and direct you based on the points raised.

> (1) broad platform support (hence the use of a [phusion/holy-build-box-64](http://phusion.github.io/holy-build-box/) build system for `linux`

This [issue]( https://github.com/conda-forge/conda-forge.github.io/issues/29 ) has basically moved in the direction of various proposals of how to move the Linux build system forward. Though there is current strategy in place, as well.

> (2) CUDA and OpenCL support (via the AMD APP SDK)...

This is under active discussion. The reason being this is tied to several issues including build system constraints, how features work, how and what of these libraries get distributed, etc.. See this [issue]( https://github.com/conda-forge/conda-forge.github.io/issues/63 ). There is a proposed example there of how we might get this to work. However, we haven't settled on something yet. 

> (3) automated builds in reproducible environments for `win`, `linux`, and `osx`.

This is all over the map. :smile: In general, we use AppVeyor (Windows), Travis CI (Mac), and Circle CI (Dockerized Linux)

If you just want to read code, we can point you there. Proper documentation isn't quite there yet. Also, there isn't one singular issue for this, but it is discussed at various points in various issues. What sort of things would you like to know?

**Comment by daler** [comment]
Hi all, checking in from bioconda. I've been poking around the conda-forge code and can't pin down where the magic is happening. Could you point to some code or to a description of what's happening to aggregate the one-recipe-per-repos?

To further the discussion, here's a description of the bioconda build system and where you can find the code. 

* Contributor submits PR
* `.travis.yaml` calls `scripts/travis-setup.sh` on OSX and Linux, which starts a docker container if linux or does the OSX setup otherwise
* `scripts/build-packages.py` is run. This does most of the work, specifically:
    * Up-to-date recipes are skipped. This lets us support 800+ recipes with very little overhead
    * Any recipes that need to be built are toposorted and added to a local file:// channel after being built. This ensures that a single PR can contain a batch of interdependent recipes and everything gets built correctly.
    * Recipe builds are treated as nosetests and therefore get the nice infrastructure that comes along with that for free (hiding stdout when not needed; keeping track of how many failures, etc)
    * If we're on a PR branch, don't upload to anaconda

The workflow is just like most anything else on github: submit a PR and wait for it to be tested. Once it passes, someone on the team merges it into master. Upon merging, travis-ci then runs again but on the master branch and this time upon completing, the built packages are uploaded to anaconda.

Aside from differences in the moving parts of the build systems, it sounds like we're all dealing with similar issues with respect to CUDA and gcc, etc. Would be nice to work out some best-practices that we could all use.

**Comment by jakirkham** [comment]
Welcome @daler.

> Could you point to some code or to a description of what's happening to aggregate the one-recipe-per-repos?

Sorry I'm not following this question. Could you please clarify what you are meaning by aggregate? It is a little unclear and I am a bit worried that there may be some misunderstanding of what is going on here. I'll try to clarify the big picture below.

> To further the discussion, here's a description of the bioconda build system and where you can find the code....

Yes, SciTools and IOOS behave in a similar manner. However, those recipes along with many from conda-recipes are being ported over here as people from those groups seem to like this model.
<br>

Just to clarify, the model for building is very different here than the many recipes in a single repo. The reasons are varied, but I think the biggest difference is it allows people to take ownership of recipes/packages that are important to them and the tools (CIs) used to test, build, and deploy. This includes making bug fixes, releases, feature support, etc. Similarly it allows relevant discussion to break along those lines. In practice, this appears to be a huge asset. However, there are plenty of other reasons for one to consider this model.

How this works:

1. Propose recipe(s) in a PR with a list of maintainers to [staged-recipes]( https://github.com/conda-forge/staged-recipes ) (here :wink:).
2. Iterate on the recipe with CIs and reviewers (at least one of them is automated :smile:).
3. PR is merged and generates repo(s) each with a single recipe and all the needed build tools.
4. Automatically, recipe maintainers are given commit privileges and control of relevant build tools.
5. Automatically, the first build is made and released (assuming all dependencies are available).
6. Do what is needed to maintain your recipe.
7. Periodically merge CI tool maintenance updates.

While understanding this infrastructure may at first seem daunting, it is actually not so bad and is not really necessary. However, if you are curious, we are more than happy to explain the details.

Maybe if you could please rephrase your question in terms of these steps, we can do a better job at answering your questions and providing you places to look for more information.
<br>

> Aside from differences in the moving parts of the build systems, it sounds like we're all dealing with similar issues with respect to CUDA and gcc, etc. Would be nice to work out some best-practices that we could all use.

Absolutely, we would be happy to point you to relevant issues where these are being discussed. Just please let me know which of these you would like to know more about.

**Comment by msarahan** [comment]
@daler, aggregation is done at the https://github.com/conda-forge/feedstocks/tree/master/feedstocks repo.  This is created with conda-smithy, particularly this module: https://github.com/conda-forge/conda-smithy/blob/master/conda_smithy/feedstocks.py

Continuum is very interested in this particular aspect (I am Continuum's representative here, though others are also involved in contributing recipes and discussing build tools).  The one-repo-per-recipe model is necessary, I think, for two reasons:

    - keep the load on the CI services small, and avoid their log size and build time limits
    - divide responsibilities and authority for each recipe with much finer granularity

The latter is the bigger issue here, since you all have had reasonable success with CI.

Continuum has started a community channel (https://anaconda.org/pycommunity), with the long-term plan to have that as a package aggregation center.  In my mind, the most important facet of this effort is to unite the recipes and have a single canonical source for each recipe.  I don't care whether it's on some project's page (i.e. matplotlib), or on conda-forge, or whatever - so long as one place is the official source, and finding that source and contributing to it is straightforward.  Conda forge is a great place to host recipes because it provides the CI of those recipes, and I like the distributed maintainer model, but I also think that hosting recipes directly at projects, and having conda-forge build from indirectly-hosted sources would be the ideal - that way the recipe would be holistically managed by the package originators.

For the pycommunity channel, we'll mirror or link packages from other channels.  In the case of multiple package sources, we haven't quite figured out how to prioritize them (activity level? origin of package?)  The hope is that rather than many organizations having to say "add our channel!" we'd instead have just one, and that one may be enabled by default for some "community edition" of miniconda/anaconda - or otherwise could be enabled with ```conda install pycommunity```

**Comment by daler** [comment]
@jakirkham and @msarahan thanks for your pointers.  One missing piece for me was that submitting a PR to `staged-recipes` triggers the CI (only travis, right?) to call `.CI/create_feedstocks`, which sets up the infrastructure, tokens etc via `conda-smithy` and transforms the repo into something similar to what's in the feedstocks repo of submodules. Is that correct?

@msarahan -- Wholeheartedly agree that a single canonical source for each recipe is critical, and that finding that source and contributing needs to be straightforward. conda-forge/conda-smithy and pycommunity look like great tools to make that happen.

**Comment by jakirkham** [comment]
> @jakirkham and @msarahan thanks for your pointers.

Glad to help, @daler. Hope it wasn't too much. Just wanted to make sure we had common context for our discussion. :smile:

> One missing piece for me was that submitting a PR to `staged-recipes` triggers the CI (only travis, right?)...

When a PR is submitted all CIs (Travis/Mac, Circle CI/Linux, AppVeyor/Windows) are run and used to attempt to build the recipe, but do not release it.

> ...to call `.CI/create_feedstocks` which sets up the infrastructure, tokens etc via `conda-smithy` and transforms the repo into something similar to what's in the feedstocks repo of submodules. Is that correct?

Once the PR is merged, a Linux job in the Travis CI build matrix does the setup for the feedstock. It goes something like this for each recipe unless otherwise specified (steps 7, 8, and 9).

1. Uses GitPython to create a new git repo on the VM with a few things committed.
     1. Cleaned up recipe
     2. Readme
     3. License
     4. `conda-forge.yml`
     5. CI files and CI support files
     6. `.gitignore`
2. Create an empty repo on GitHub (via PyGithub).
3. Push to the GitHub feedstock repo.
4. Configure all of the CIs for the repo on GitHub.
5. Commit info to CI configuration files for uploading binaries to Anaconda.org.
6. Push to the GitHub feedstock repo.
7. Delete all the recipes from the staged-recipes and commit.
8. Push to the GitHub staged-recipe repo.
9. Trigger a global feedstock update.

As you have mentioned, this all basically happens through `conda-smithy`. However, there is some code that lives here for that purpose too. Take a look at this [log]( https://travis-ci.org/conda-forge/staged-recipes/jobs/121962528 ) for `configparser` and `entrypoints` to get a better idea.

After generating a feedstock, a global feedstock update is run. It is pretty simple. It updates the [feedstocks]( https://github.com/conda-forge/feedstocks ) with the latest commit of each feedstock on `master` at conda-forge. It also updates the [listing]( https://conda-forge.github.io/feedstocks.html ). However, changes may not be reflected in the listing immediately even if the changes have been made to the HTML source code due to how GitHub caches GitHub Pages.

**Comment by daler** [comment]
Perfect, these were just the kinds of details I was looking for. Thanks. Hopefully it can help get others up to speed as they join the discussion as well.

**Comment by johanneskoester** [comment]
Hi guys,
thanks for initiating this. It is very interesting to exchange ideas of how to build.
I have two questions:
1. Have you every considered using the anaconda build service? I recently had a look at it, and it seems to me centered on packages instead of repositories/organizations, which is kind of unfortunate, because it needs to be set up for each package, right?
2. With your conda-forge model, how do you deal with dependencies between recipes?

**Comment by msarahan** [comment]
> Have you every considered using the anaconda build service? I recently had a look at it, and it seems to me centered on packages instead of repositories/organizations, which is kind of unfortunate, because it needs to be set up for each package, right?

Yes, especially for Windows builds.  Mapping conda-forge's model to Anaconda.org should be OK - the organization would be conda-forge, and each package would be a different build.  Maybe I'm missing how this is different from the other CI services?  Anyway, the hangup has been that anaconda.org has some kinks that need to be worked out.

> With your conda-forge model, how do you deal with dependencies between recipes?

ATM, I think the answer is "we don't."  There has been discussion about coming up with networkx-driven guidance of what recipes to work on next, but that has been for human consumption more than automated buildout of dependency trees.  Before getting involved in conda-forge, Continuum developed a build script that also uses networkx, and builds out these trees.  That code assumes a single folder of packages, which can be created from conda-forge using conda-smithy.  The dependency building code is part of ProtoCI: https://github.com/ContinuumIO/ProtoCI/blob/master/protoci/build2.py

**Comment by johanneskoester** [comment]
Thanks for the clarification.
My point is the following: if the anaconda build service could be setup per repository and not per package, CI job limits are no reason any more to have separate repositories per recipe, right?

**Comment by msarahan** [comment]
I think separate repos per recipe are still a good thing, because it gives you complete control over who has permission to accept changes to a recipe.  I don't know how we'd do that with many recipes under one umbrella.

**Comment by jakirkham** [comment]
>  Before getting involved in conda-forge, Continuum developed a build script that also uses networkx, and builds out these trees. That code assumes a single folder of packages, which can be created from conda-forge using conda-smithy. The dependency building code is part of ProtoCI: https://github.com/ContinuumIO/ProtoCI/blob/master/protoci/build2.py

Would this work on the `feedstocks` repo possibly with some tweaks? This might be a good way to get things going and it would also avoid having several scripts created here that kind of do something like this. Thoughts?

**Comment by msarahan** [comment]
Sure, I think so.  It would need to be adapted to look into the nested recipes folder, but I think otherwise, it would work fine.  It may also have trouble with jinja vs. static version numbers - but again, that's tractable.

**Comment by johanneskoester** [comment]
@msarahan I agree, this is in general a nice advantage. I asked, because the situation if different for bioconda. There, we have a rather controlled collaborative community, and it is much more convenient to have all recipes in one repository (e.g. for toposorting builds).

**Comment by msarahan** [comment]
Yeah, the one thing we don't have figured out well yet is how to edit multiple recipes at once.  For aggregating them and building them as a set, I think conda-smithy + ProtoCI abstract away the difficulties with one repo per recipe.

**Comment by johanneskoester** [comment]
But if you build them as a set, you have the problem with job limits in the CI again, haven't you?

**Comment by jakirkham** [comment]
Yeah, I figure the nested directory structure needs to be addressed. Otherwise adding jinja template handing is probably valuable no matter where it is used, no?

**Comment by msarahan** [comment]
> adding jinja template handing is probably valuable no matter where it is used, no?

Absolutely.  In case you missed it, @pelson has a nice snippet at https://github.com/conda-forge/shapely-feedstock/issues/5#issuecomment-208377012

**Comment by jakirkham** [comment]
> But if you build them as a set, you have the problem with job limits in the CI again, haven't you?

Well, one could consider some sort of debouncing to handle this. Namely even though one has made the change together and one is submitting them all ultimately, we manage submissions/builds somehow so that they are staggered. This will likely require some thought, but it is useful for some workflows with the recipes.

**Comment by msarahan** [comment]
> But if you build them as a set, you have the problem with job limits in the CI again, haven't you?

With anaconda.org, we don't have artificial limits.  There are still strange practical limits - like logs that get too large end up making web servers time out.  These are tractable problems.

**Comment by jakirkham** [comment]
Interesting, thanks for the link. I'll take a closer look.

**Comment by johanneskoester** [comment]
@msarahan, I know, you don't have these limits, but my understanding was that anaconda.org cannot out of the box build recipes as a set, right? You have to register an individual trigger for each of them? And then, their order of execution is no longer determined, and they can't depend on each other. Or am I missing something here?

**Comment by msarahan** [comment]
@johanneskoester there would need to be some intermediate representation as a collection of recipes.  Then that ProtoCI tool would then be able to build things that changed.  It is written to build packages based on which packages are affected by a git commit.  Here, obviously only one recipe could trigger, rather than many changing at once.  That does not affect its ability to build requirements dependencies, though - and they'll be built in topological order.

**Comment by johanneskoester** [comment]
@msarahan, maybe this is not the right thread (I don't want to bother the rest with my detail questions here, so feel free to stop me if you feel this becomes off-topic).
Ok, but protoCI has to run e.g. on travis, right? That means, even if protoCI triggers builds on anaconda.org in the right order, travis would still need to wait on the results, in order to be able to report back to github? Which would result in the same timeout issues? Sorry if I misunderstand something here.

**Comment by msarahan** [comment]
ProtoCI was designed to run on anaconda.org.  If anyone else wants to get it to run on Travis, that's cool, but that wasn't what it was written for.  It would not be triggered by Travis or any other build - rather, it would be a new CI service in addition to or instead of the existing CI services.

**Comment by johanneskoester** [comment]
Great! Now it makes sense, sorry I did not know that. So, is there any documentation on how we could set up ProtoCI for bioconda? We already have a repository with multiple recipes in place. Or is that not possible yet?

**Comment by msarahan** [comment]
I'm not completely sure how possible it is.  I have been involved, but not the one doing most of the real work.  It is live on conda-recipes, and you should start with the binstar.yml there as an example, but you'll have to tweak it for your build queue:

https://github.com/conda/conda-recipes/blob/master/.binstar.yml

In short, protoci should be installed on the build workers.  It is on ours, but I can't speak for how you have your queue set up.  Your build script should just call the protoci-difference-build entry point.

**Comment by johanneskoester** [comment]
Thanks Mike, that's good news!

**Comment by johanneskoester** [comment]
A question regarding conda-forge: On which linux system/libc version do you build?

**Comment by msarahan** [comment]
That's sort of in flux.  See https://github.com/conda-forge/conda-forge.github.io/issues/29

I think the current one is this: https://github.com/pelson/Obvious-CI/blob/master/obvious-ci.docker/linux64_obvci/Dockerfile

**Comment by bgruening** [comment]
Thanks for this healthy discussion, exactly what I wanted to trigger :)
One additional question from me. Does conda-forge has any other channels (`default` I assume) activated during build time?

**Comment by ocefpaf** [comment]
> Does conda-forge has any other channels (`default` I assume) activated during build time?

Nope. Only the `default` channel.

**Comment by patricksnape** [comment]
Just to say that I'm in favour of the single repos as it currently is in conda-forge/feedstocks. Although I didn't go as ambitious as the bioconda/ioos/scitools/omnia crowd I've also been maintaining a set of recipes that we needed for our project, menpo. Most importantly, I've been really trying to drive Windows support because so many people in Computer Vision still use Windows due to historical (Matlab) reasons. So I'm usually keen to try and help with upstream support for Windows (as well as @msarahan who has been the real Windows champion). 

I'm very interested in the CUDA/OpenCL builds that you guys seem to have. I wonder if we could become to the goto place to pickup projects like Theano for deep learning?

**Comment by pelson** [comment]
> Just to say that I'm in favour of the single repos as it currently is in conda-forge/feedstocks

Thanks for the honesty. Technically conda-forge/staged-recipes is a many recipe repository too, it is just we've automated it so that recipes immediately get deleted and added to their own repository on merge. 😉 

With that in mind, you may be aware of [conda-build-all](https://github.com/SciTools/conda-build-all) which is the tool we use in this repo to build all recipes in a single repo. It (and its predecessor ObviousCI) was the tool we used in IOOS and SciTools (amongst others) to build and upload to our respective channels. Because of staged-recipes dependence upon it, we are going to need to continue to maintain that capability, so if your looking for shared tooling for the single-repo many-recipe usecase, you might want to take a look.

As you've highlighted, even if you don't favour the approach we have taken at conda-forge, there is still huge potential for us to collaborate so that we can collectively package in a consistent and coherent way. Your input so far has been exceptionally valuable, and long may it continue! 👍

**Comment by bgruening** [comment]
>
As you've highlighted, even if you don't favour the approach we have taken at conda-forge, there is still huge potential for us to collaborate so that we can collectively package in a consistent and coherent way. Your input so far has been exceptionally valuable, and long may it continue! :+1:

I would like to highlight this and raise 3 points to start a closer cooperation.

1. Develop a common build-infrastructure for all conda community projects. This includes best-practise guidelines of dependencies, e.g. specifying gcc/llvm, is zlib as dependency needed ...
2. Enable more channels by default. A good start would be if bioconda could activate conda-forge and vice versa as default channel during build.
3. Develop recommendations about the correct channel for a specific package. For example libraries should always be stored in conda-forge. Bioinformatic software in bioconda and cheminformatics in omnia - as long as we still want to have separate repositories.

**Comment by johanneskoester** [comment]
@bgruening this sounds very reasonable. General purpose libraries can go into conda-forge, but I can also imagine to just pass them over to Continuum. I am also not quite sure about whether to prefer conda-recipes or conda-forge for that... Can I expect that conda-forge PRs are handled faster than on conda-recipes?

**Comment by bgruening** [comment]
> Can I expect that conda-forge PRs are handled faster than on conda-recipes?

This was my expectation. One huge benefit of the bioconda model is that if you need a recipe in 30min you get get those. conda-forge is hopefully way faster than conda-recipes :)

**Comment by pelson** [comment]
>> Can I expect that conda-forge PRs are handled faster than on conda-recipes?

> This was my expectation. One huge benefit of the bioconda model is that if you need a recipe in 30min you get get those. conda-forge is hopefully way faster than conda-recipes :)

If you are proposing a new recipe, then you have to wait on one of the conda-forge "staged-recipes" maintainers.

If you are proposing an update to an existing recipe, then you have to wait on one of the feedstock maintainers (as listed in the recipe/meta.yaml).

If you are proposing an update to an existing recipe for which you are a maintainer, you are waiting on yourself (and maybe the CI to finish).

In general I'm keen to be very open about membership of the "staged-recipes" group. The important qualities of a maintainer in that team is an eye for detail, a feeling for what is "maintainable", and shed loads of experience of reading and writing conda recipes. I suspect most people in this thread meet that criteria, and after proposing just 3 or 4 PRs which merge smoothly, I'd be happy to say that was a good candidate for membership (though with all the noise on conda-forge, it is probably necessary to ask, rather than have it suggested).

**Comment by 183amir** [comment]
Hey guys, I have ported a couple of my own recipes from [bob.conda](https://github.com/bioidiap/bob.conda) to `conda-forge` and I have used [this script](https://gist.github.com/183amir/0e868e045dde66d671f16870f6cf184f) to port things over. It is not well written and I am sure you can write better scripts but as @jakirkham mentioned maybe sharing it with you guys could help you automate your porting process.

**Comment by jakirkham** [comment]
Thanks for sharing this, @183amir. :smile:

**Comment by 183amir** [comment]
I guess the most important part is that I load recipes with ruamel.yaml and take the `example` recipe as base and [update it](https://gist.github.com/183amir/0e868e045dde66d671f16870f6cf184f#file-patch_meta_forge-py-L166) with my own recipe.

**Comment by johanneskoester** [comment]
At bioconda, we are evaluating if it makes sense to specify the compiler in the recipe. Our feeling is that this provides some advantages, e.g.

* the recipe can define constraints on the minimum compiler version
* conflicts can be detected via depending on libgcc at runtime
* explicit is better than implicit

We are unsure though about osx (since clang seems to be used there by e.g. the default channel).
What are your thoughts on this? It looks like no conda-forge recipe is currently depending on gcc.

**Comment by jakirkham** [comment]
Hi @johanneskoester, sorry your comment got buried in notifications I am afraid and am now discovering while going back through some things.

Generally, our feeling here is we want to move away from using a `gcc` package. We have been making steps in that direction. In particular, we now use CentOS 6 with devtoolset-2 for nearly all of our building on Linux. On Mac, we require 10.7 where using the system compiler (`clang`) has C++11 support. This largely meets our needs at present. There are a few exceptions when dealing with OpenMP and/or Fortran. Though we may re-evaluate our strategy here in the future. See this issue ( https://github.com/conda-forge/conda-forge.github.io/issues/29 ) for more details on various proposals.

While it is a nice idea in theory to use a compiler package, in practice this doesn't fair so well. One reason is we can't make any guarantees about GLIBC compatibility on Linux as the compiler could be used to build anywhere. So, we have opted to work on proper docker containers that include a compiler in them. Another reason this doesn't work well is that if we run into an issue with the compiler package (as we did recently), we are largely incapable of fixing it due its long build time that exceeds CI limits. As a result, we can at best use kludgy hacks to try and solve the problem. In the worst case, we find ourselves crippled.

Maybe a better long term strategy that provides the same guarantees without the same issues would be to create a pseudo compiler package. This package could be used to verify that a compatible compiler can be found. Additionally, this package could be used to perform some sort of configuration to ensure the compiler it found is used. This would allow proper constraints in an explicit manner, but avoid the pains associated with the packaged compiler.

**Comment by johanneskoester** [comment]
Thanks for the answer! We also used devtools-2 before, but unfortunately we went in exactly the opposite direction, requiring the gcc package for all recipes that compile something. Regarding the gcc issues: They appear to be in the conda-forge gcc package, right? Why do you shadow the gcc package from the default channel at all? That one also should not have the CI issues, because Continuum builds on anaconda.org, right?

**Comment by jakirkham** [comment]
> Thanks for the answer! We also used devtools-2 before, but unfortunately we went in exactly the opposite direction, requiring the gcc package for all recipes that compile something.

I see. Well, we are partially following @msarahan's lead here. Though maybe a bit slower than he would like. He has made a strong case for not using the packaged gcc.

> Regarding the gcc issues: They appear to be in the conda-forge gcc package, right?

Nope. We tried to package it because of the problems we had with it, but we couldn't.

> Why do you shadow the gcc package from the default channel at all?

As stated before, we don't.

> That one also should not have the CI issues, because Continuum builds on anaconda.org, right?

Not sure I follow this question. Could you please clarify what you mean here?

**Comment by msarahan** [comment]
TL;DR: I wish we could just package the compiler.  I don't think we can,
because we need to understand and control compatibility as best we can.

It's fine for us to use a packaged gcc.  What's not fine is for someone who
does not understand the compatibility implications to build packages that
then don't work on some older systems, raising confusing support issues.
If we either used a different c runtime (musl? Lots of work, IMHO) or track
runtime needs and have conda be smarter about them (we're working on this),
then I don't care what compiler anyone uses, aside from rarifying the
available packages for particular platforms.

I had hoped to drive adoption of a common build platform so that the
collection of packages would not vary in regards to support platforms.
Continuum is sometimes using the old standard, CentOS5, GCC 4.1 or 4.4, and
sometimes using the Docker image that I created with CentOS 5 and GCC 5.2.
The numba team modernized slightly to CentOS 6 for NVidia's tools.  They
are using devtoolset-4 (GCC 5.2) rather than a GCC that we compile, because
they didn't want the overhead of compiling GCC.

A docker image of some sort is the easiest way to nail down the glibc
version that we support.  This ends up being the most critical factor in
backwards support.  For forwards support, we either need to depend on
devtoolset's partial static linking, which @tkelman and @mingwandroid seem
to be skeptical of, or we need to ship runtimes that are always at least as
new as the systems we hope to support.  If we don't keep up with this,
people will see errors like those being reported for libgfortran (which is
version 3, but with an old fortran API - code built with a newer fortran
compiler doesn't find what it needs):
https://github.com/ContinuumIO/anaconda-issues/issues/686

**Comment by tkelman** [comment]
> musl? Lots of work, IMHO

GCC 6 now has a `-mmusl` flag, which I admit I haven't played with yet. Check Alpine Linux's packages, they tend to need a lot of patches around glibc assumptions all over the place, the patches needed for big things like LLVM are especially hacky and hard to get upstreamed. Hopefully musl distros will become more popular and cleaner versions of these patches will gradually all get merged.

> devtoolset's partial static linking, which @tkelman and @mingwandroid seem to be skeptical of

If I saw devtoolset correctly statically link things like an openblas build that includes lapack support, or arpack, or mumps, etc then I'd be more satisfied with it. When we tried using it to build Julia all of the Fortran dependencies still had shared-library dependencies on libgfortran, but to the really old libgfortran from the centos build system, which we don't want to ship as it has the above issue.

The folks responsible for manylinux wheels have been advocating renaming the runtime libraries to avoid conflicts, this needs some patchelf fixups and might get a bit messy but we'll see how they get on with that approach.

My GCC-6-on-CentOS-6 docker image did eventually go through with Quay's autobuilder, it takes a while but I could probably cut the build time down by only doing one phase of bootstrap instead of the full default gcc build process. It's at https://github.com/tkelman/c6g6 (with various branches where I'm testing out building on different CI services and building Julia as a proof of functionality).

**Comment by msarahan** [comment]
@tkelman

> When we tried using it to build Julia all of the Fortran dependencies
still had shared-library dependencies on libgfortran, but to the really old
libgfortran from the centos build system, which we don't want to ship as it
has the above issue.

I think this is by design - it's not a full static link by default.  It
includes only the newer parts, and assumes installation of the older
libraries.  If you were trying a full static link, and it didn't work, then
that's certainly dysfunctional.

I'm closely following the manylinux renaming effort, and may soon try to
adapt their tools to conda-build.

Great to hear that there is a CI platform that actually build GCC, though.
That's quite an accomplishment.  Kudos!

**Comment by mingwandroid** [comment]
@msarahan, I agree that the manylinux renaming effort should be integrated into conda-build. Though it also should be renamed considering https://github.com/njsmith/redll

**Comment by tkelman** [comment]
I think linuxbrew actually packages its own recent glibc, which might be another option to avoid having to control the build system so precisely.

**Comment by jakirkham** [comment]
> I think linuxbrew actually packages its own recent glibc, which might be another option to avoid having to control the build system so precisely.

Really? I have gotten the impression that it is a bad plan to package glibc. Why is this ok?

**Comment by Lnaden** [comment]
Hello! I am wanting to let you all know that the `omnia` community has decided to start our migration over to `conda-forge` and I was hoping to revive the discussion or at least get help on how we can handle some packages which would require things beyond the current scope of the `linux-anvil` 

Right now our plan is to handle the migration in stages, with the end goal of fully moving off the `omnia` conda channel.

1. [switch](https://github.com/omnia-md/conda-recipes/pull/761) from our Docker file running from the phusion/holy-build-box to a [modified version](https://github.com/omnia-md/omnia-linux-anvil/blob/texlive-amd30-cuda80/Dockerfile) of conda-forge's linux anvil.
2. Have our individual package maintainers start moving their packages to conda-forge such that they no longer depend on the `omnia` channel.
3. Move packages which require additional software not currently available through conda(-forge) over
4. Have all of omnia's packages on conda-forge.

With regards to the 3rd item on the list, I was hoping to get a discussion as to how to handle the additional tools some of our packages would need into conda-forge. 

Here are the capabilities as I see them we would need and would like to discuss:
1. Ability to install extra yum packages
    * I think this already exists through the `yum_requirements.txt` file, but I can find very little documentation on how this should work
    * It would be nice to get packages off the EPEL as well
2. Installing TeX packages, but that might already be possible through the texlive-core package
3. Installing GPU libraries to build against such as nVidia's CUDA and AMD's OpenCL SDK
    * cross-ref https://github.com/conda-forge/conda-forge.github.io/issues/63

I would also like to discuss the possibility of supporting extended Docker images which is how we are currently handling step 1 of our migration process, although this would have to be highly regulated to ensure compatibility across conda-forge.

There is alot to digest here, so I'm more than happy to split off parts of this to other issues as needed. 

Pinging @jchodera

**Comment by bgruening** [comment]
@Lnaden this will made my day! Awesome!
Regarding 3) do you have a list of packages that is needed?

A few of the omnia packages are already in BioConda which is using the same Docker Image as conda-forge, so it might be easier to port the BioConda package or use BioConda as well.

**Comment by ocefpaf** [comment]
@Lnaden that is really good news. We are going to hold meeting tomorrow see

https://gitter.im/conda-forge/conda-forge.github.io?at=58fe073ccfec9192726d5141

and 

https://conda-forge.hackpad.com/conda-forge-meetings-2YkV96cvxPG

for the details. If you are available at that time try to participate.

**Comment by Lnaden** [comment]
@bgruening It may take a while to migrate all the packages since we will be relying on individual package developers to move their own packages, but the first step will be to force them to add `conda-forge` as a channel. ;) I personally will also be very happy to not have to maintain our own `conda-build-all` script anymore.

To see everything we are doing differently, you can see the [extended linux-anvil](https://github.com/omnia-md/omnia-linux-anvil/blob/texlive-amd30-cuda80/Dockerfile) we are preparing to use [not merged in yet](https://github.com/omnia-md/conda-recipes/pull/761). Its a bit crude since I know several packages are already installed but here is the list of things we need as of now:

**yum**:
perl dkms libvdpau git wget libXext libSM libXrender groff dkms libvdpau

of which I know libXext, libSM, libXrender, and groff are already installed based on our Docker build log. dkms and libvdpau are for the CUDA toolkits

**texlive**
The core + 21 extra packages, not going to list them here. Several are to ensure OpenMM can build docs against Sphinx 1.5.1 which stopped shipping a number of their own latex packages

**GPU Related**
* AMD SDK 3.0
* CUDA Toolkit 8.0

Not every package however needs all of those additions. The big one is OpenMM (openmm.org) which requires the GPU files, lots of the TeX, and some of the yum packages. Because we have a number of tools which rely on OpenMM though, we won't be able to move a number of the packages until we can move it first. 

The packages which do not depend on OpenMM should be much easier to move over, and some packages such as mdtraj moved quite a while go.

**Comment by Lnaden** [comment]
@ocefpaf I should be able to attend the meeting. I also want to ping @peastman @mpharrigan and @rmcgibbo who may also be interested in being on that meeting if they can (5-6 PM UTC)

**Comment by jchodera** [comment]
I suspect we may be able to avoid requiring `texlive` if we can get the [`texlive-core`](https://anaconda.org/conda-forge/texlive-core) conda-forge package to work, but we had some issues getting `tlmgr` to correctly install the needed tex packages.

**Comment by bgruening** [comment]
@Lnaden any reason you need those libraries inside the container, from a short look it seems that many of your libs already package, like the X.org stack, wget, perl - so no need to put this into the container.

**Comment by Lnaden** [comment]
@bgruening The split Docker image is a port from the CentOS 5-based Docker file [here](https://github.com/omnia-md/omnia-build-box/blob/cuda80-amd30/Dockerfile) and [here](https://github.com/omnia-md/omnia-build-box/blob/cuda80-amd30-clang38/Dockerfile) and I have not refined it yet. 

I fully acknowledge there are packages which are already part of the base conda-forge `linux-anvil` that could be removed, but there are other packages such as dkms and libvdpau which we would still need. I'm also not sure about how much of the CUDA downloads we actually need, what we have now may be overkill, but we would need at least some of them, and those are not available through `yum`.

**Comment by jakirkham** [comment]
> I suspect we may be able to avoid requiring `texlive` if we can get the [`texlive-core`](https://anaconda.org/conda-forge/texlive-core) conda-forge package to work, but we had some issues getting `tlmgr` to correctly install the needed tex packages.

I tried some time ago to get this to install `tlmgr`. FWIR it doesn't get installed with the way we are installing TexLive. It seems most distros prefer not to include `tlmgr` and instead package everything themselves. TBH this sort of makes sense as it is possible to end up with broken things when two package managers get involved. For instance, suppose `tlmgr` downloaded some packages that require some image libraries. How does it resolve this? Does it use system ones from some standard location? What if it can't find them? Does it try to install the libraries itself? How do we teach it where `conda`'s libraries are? There may be sensible solutions to this. Then again, we may end up hacking `tlmgr` and every package it provides to the point where we are better off packaging these ourselves as it will be less work. Something to consider.

**Comment by jakirkham** [comment]
> Ability to install extra yum packages

Yeah, just add a `yum_requirements.txt` file in the recipe directory. Here's an [example]( https://github.com/conda-forge/qtconsole-feedstock/blob/b4be35927c1593116a5565d103289005049391d5/recipe/yum_requirements.txt ). Though if anything else needs a package that uses a `yum_requirements.txt` file, it will need to include all the content from that package's `yum_requirements.txt` recipe and anything else it might need. Beyond that, everything is tooled to correctly handle this case.

> It would be nice to get packages off the EPEL as well

We should talk more about this in a separate issue. Maybe on the webpage repo. It would be good to hear what packages are required from this source. Generally we have been moving to use `yum` less, not more. So we would need to understand why these can't otherwise be packaged.

**Comment by Lnaden** [comment]
> most distros prefer not to include tlmgr and instead package everything themselves.

I would love if all distros actually packaged their own tex packages themselves, and I would very much prefer not to rely on a conda-based TeX install. From my very brief testing of the `texlive-core` package on conda-forge, the `tlmgr` it installs fails horribly if you have a global tex installed on your computer since the libraries are mis-matched. One problem we had recently was upgrading Sphinx from 1.3.1 to 1.5+ where they actually stopped shipping several tex packages so we had to include them.

The only thing I think the packages in omnia use TeX for is building the docs to ship with the build. Maybe an option is to make TeXLive available on the image so packages can add `tlmgr` commands to their `build.sh` scripts and ship their compiled docs. If users want to build the docs themselves, they will have to rely on their own local TexLive. Just a thought.

> Yeah, just add a yum_requirements.txt file in the recipe directory.

I have seen several packages do this. My larger concern is the lack of documentation on this feature, the best I could find was mention of it [on conda-smithy's page of the docs](http://conda-forge.github.io/docs/conda_smithy.html).

> We should talk more about this in a separate issue. Maybe on the webpage repo.

Happy to. I know at least DKMS and libvdpau from EPEL is needed to make the CUDA toolkits work correctly, so if we can solve the GPU libraries instead, that may remove the need to get EPEL and allow us to shift away from heavy `yum` usage.

**Comment by jakirkham** [comment]
> My larger concern is the lack of documentation on this feature...

Yeah, documentation is a weak point for conda-forge still. That said, @pelson and I wrote the `yum_requirements.txt` functionality and there are many important recipes (e.g. `qt`) that depend on it. So if you are worried about it doing something surprising or disappearing, I don't think that should be a concern. That said, maybe worth adding an issue at the webpage repo to document its behavior.

**Comment by jakirkham** [comment]
> Installing GPU libraries to build against such as nVidia's CUDA and AMD's OpenCL SDK 

Basically this has been a problem of licensing. The same issue occurs with MKL for instance. It's been awhile since I've looked into it.

I know that Continuum was able to get permission to build and distribute CUDA and cuDNN libraries. Though they are using CUDA 7.5 not 8 currently. If there is only one or two low-level packages like that, we could request they get build and added to `defaults`. That said, we would still need the recipes for them.

Alternatively we could try emailing NVIDIA directly and asking how to use their toolchain in CIs and how to distribute the CUDA libraries. Considering the size and audience of conda-forge, I think NVIDIA would be interested in a mutually beneficial solution. Would you have interest in writing such a letter? We could put this in a Google Doc to get feedback from others.

Haven't looked into AMD's OpenCL SDK yet. Though am getting the impression that sending an email their way wouldn't hurt either.

In any event, let's move this discussion over to issue ( https://github.com/conda-forge/conda-forge.github.io/issues/63 ).

**Comment by jjhelmus** [comment]
Continuum is distributing the CUDA and cuDNN libraries and headers in the `cudatoolkit` and `cudnn` packages.  To build packages which compile against these libraries you need to install the CUDA command line tools (nvcc, etc) on the bare metal/VM/Docker container as well as have a host machine with the correct NVidia drivers installed.  AFAIK, on CentOS the driver require DKMS which is only supported by installing EPEL packges.  I do not believe these tools can be distributed, although NVidia does provide [docker images with these tools installed](https://hub.docker.com/r/nvidia/cuda/) but it is unclear to me if new images created from these can be re-distributed.

**Comment by jchodera** [comment]
To answer @bgruening's question:
> @Lnaden any reason you need those libraries inside the container, from a short look it seems that many of your libs already package, like the X.org stack, wget, perl - so no need to put this into the container.

When we set up our [`omnia-build-box`](https://github.com/omnia-md/omnia-build-box/blob/cuda80-amd30/Dockerfile) docker image, we found those `yum` dependencies were required in order to install the CUDA toolkit and libraries, as @jjhelmus suggests. I don't think we'd be able to migrate most of those dependencies out to packages.

> Alternatively we could try emailing NVIDIA directly and asking how to use their toolchain in CIs and how to distribute the CUDA libraries. Considering the size and audience of conda-forge, I think NVIDIA would be interested in a mutually beneficial solution. Would you have interest in writing such a letter? We could put this in a Google Doc to get feedback from others.

@jakirkham : The first thing I would do would be to bring in [Mark Berger](https://www.linkedin.com/in/1markberger), the Senior Alliance Manager at NVIDIA, to help connect us directly with someone at NVIDIA who can help negotiate any technical/licensing issues. He is incredibly supportive of building an ecosystem for scientific computing that can exploit NVIDIA hardware (that's his job!), and making CUDA-enabled packages more accessible could only further that goal. We can help initiate that contact by email on our side as needed---@lnaden can help coordinate.

**Comment by jakirkham** [comment]
> To build packages which compile against these libraries you need to install the CUDA command line tools (nvcc, etc) on the bare metal/VM/Docker container as well as have a host machine with the correct NVidia drivers installed. AFAIK, on CentOS the driver require DKMS which is only supported by installing EPEL packges.

Thanks for the details @jjhelmus. Maybe we should come up with some scripts to automate this process on the CIs. If they just live in CI builds, that should minimize the chances of accidentally distributing these.

>  The first thing I would do would be to bring in Mark Berger...

That's fantastic @jchodera. Please do.

**Comment by jjhelmus** [comment]
Do any of the CI providers have nodes with NVIDIA GPUs?  I do not know if these would be strictly needed to compile GPU packages but they would be needed to properly test the packages.

**Comment by Lnaden** [comment]
That is one problem we have with OpenMM right now, our builds on Travis and Appveyor cannot test the GPU components correctly. Our current process involves building and uploading OpenMM on a `dev` label to the `omnia` conda channel, testing the packages locally on machines with GPUs, then moving them to the `main` tag when it checks out. Not the most automated process.

IIRC you can set up local Jenkins tests to handle the GPU, but I think that requires local physical boxes you can set up with private access, and clearly not a viable solution for conda-forge

**Comment by johanneskoester** [comment]
At Bioconda, we are currently experimenting with [Buildkite](https://buildkite.com). Basically, Buildkite provides an interface and management framework for CI agents that are deployed to local machines. This has two major advantages

1. Full control over the used hardware (including GPUs)
2. No build time limits or bottlenecks, since the build system can be easily scaled by connecting new machines.

With such an approach, security of the used systems becomes important of course. For example, they should always be in some kind of DMZ, and build jobs should be executed in a volatile, containerized or virtualized environment.

**Comment by jakirkham** [comment]
How does Buildkite compare to say [Concourse CI]( https://concourse.ci/ ), @johanneskoester? Asking because @msarahan has been working on using Concourse CI to build `conda` packages.

ref: https://github.com/conda/conda-concourse-ci

**Comment by msarahan** [comment]
On the surface, they look much the same.  The main reason I chose Concourse
was the ability to dynamically create batch jobs.  We have the ability to
monitor a git repo of many recipes (or subrepos), and build all the ones
that have changed.  Having the aggregate view be the working set of recipes
also means we can do neat things like testing n steps downstream from a
newly built package to make sure things still work.

**Comment by johanneskoester** [comment]
Indeed, looks similar. With buildkite, dynamic batch jobs are possible as well (via their agent tool). I don't know if concouse makes this easier though.

Note that we have stopped our experiments with builkite though. We mainly where interested because of the build time limits on Travis, but we found a way to circumvent that nicely within Travis CI. For large bulk updates, we have simply one fixed branch. When we update our global pinnings on that branch, all affected recipes are rebuilt. Whenever a built succeeds, the package is uploaded to anaconda. Since hundreds of recipes can be affected, the build time limits can be exceeded. However, since we upload immediately after success, no recipe has to be built twice. Instead, we can simply fix all failed recipes, and trigger a rebuild by pushing the fixes. Now, in the next iteration, new recipes will be built in addition to the fixed ones, and so on. By this, we can simply reduce the number of remaining recipes to zero with a couple of iterations on the bulk branch. This worked really well when recently updating Python to 3.6 and R to 3.3.2. And we have all recipes in a single repository, so it is easy to keep track of the dependency DAG in order to build in the correct order.

**Comment by jakirkham** [comment]
Closing as I think this has served its purpose. Happy to continue relevant discussions in new issues.
